data/code/lib.c

```diff
@@ -589,3 +593,732 @@ void check_lap_ex()
 #endif
 }
 #endif
+
+#define EXTENDED_TERRAIN_BOUNCE_FLAG 0x1
+#define EXTENDED_TERRAIN_BOUNCE_LIFTOFF_FLAG 0x2
+
+#define EXTENDED_TERRAIN_BOUNCY 0xB0
+
+// The importance of this remaining in its own section is that its functionality will be needed for
+// any future custom terrain types added, within the Extended Terrain Types patch or otherwise.
+#if EXTENDED_TERRAIN_TYPES
+
+typedef struct
+{
+    int point_indexes[3];
+    float unknown;
+    short normal[3];
+    char terrain_type;
+    char terrain_type_part_two;
+    char min_max_table;
+    char camera_code;
+    short neighbor_triangles[3];
+    int splash_hash;
+    // splash_hash is actually a 4-byte structure.
+
+} CollisionTriangle;
+
+struct CrsGround
+{
+    char unknown[0x20];
+    struct CollisionTriangle* col_triangle;
+};
+
+// Check against all of the custom material flags enabled by the patch.
+int is_extended_terrain_type(const char terrain_type)
+{
+    return terrain_type == EXTENDED_TERRAIN_BOUNCY;
+}
+
+int should_return_fake_code(const CollisionTriangle* triangle)
+{
+    if (triangle)
+    {
+        return is_extended_terrain_type(triangle->terrain_type);
+    }
+
+    return 0;
+}
+
+// Game will search for a Splash object due to the material hash being used.
+// This nullifies that behaviour.
+struct CollisionTriangle* get_splash_code_inline()
+{
+    register const struct CrsGround* const ground asm("r3");
+    if (should_return_fake_code(ground->col_triangle))
+    {
+        return 0;
+    }
+
+    return ground->col_triangle;
+}
+
+// Game does not want material flags it does not recognize to allow for items to collide with them.
+// This allows items to sit on custom materials as one would ordinarily expect.
+void is_item_inval_ground_hijack()
+{
+    register const struct CrsGround* const ground asm("r3");
+    if (should_return_fake_code(ground->col_triangle) != 1)
```

Comment:
In C, I'm always scared of these functions potentially returning `2` or other value that should still be resolved to `true`. How about:

```diff
- if (should_return_fake_code(ground->col_triangle) != 1)
+ if (!should_return_fake_code(ground->col_triangle))
```
(There may be some more cases like this one.)

Reply:
It's only this way because I was scared that treating it like an ordinary `bool` would cause some problem. I guess I was just being paranoid;  I would ordinarily have done this conditional the way you are suggesting.

Reply:
When I went through to address ` != true` conditionals, I also thoughtlessly changed many `== true` without thinking about why I shouldn't do this. I've reverted this change with its own commit, but I also understand the history of pointless changes is not preferrable.